pyro/infer/mcmc/api.py

```diff
@@ -0,0 +1,322 @@
+                    # Exception trace is already logged by worker.
+                    raise val
+                if val is not None:
+                    yield val, chain_id
+                else:
+                    active_workers -= 1
+        finally:
+            self.terminate()
+
+
+class MCMC(object):
+    """
+    Wrapper class for Markov Chain Monte Carlo algorithms. Specific MCMC algorithms
+    are TraceKernel instances and need to be supplied as a ``kernel`` argument
+    to the constructor.
+
+    .. note:: The case of `num_chains > 1` uses python multiprocessing to
+        run parallel chains in multiple processes. This goes with the usual
+        caveats around multiprocessing in python, e.g. the model used to
+        initialize the ``kernel`` must be serializable via `pickle`, and the
+        performance / constraints will be platform dependent (e.g. only
+        the "spawn" context is available in Windows). This has also not
+        been extensively tested on the Windows platform.
+
+    :param kernel: An instance of the ``TraceKernel`` class, which when
+        given an execution trace returns another sample trace from the target
+        (posterior) distribution.
+    :param int num_samples: The number of samples that need to be generated,
+        excluding the samples discarded during the warmup phase.
+    :param int warmup_steps: Number of warmup iterations. The samples generated
+        during the warmup phase are discarded. If not provided, default is
+        half of `num_samples`.
+    :param int num_chains: Number of MCMC chains to run in parallel. Depending on
+        whether `num_chains` is 1 or more than 1, this class internally dispatches
+        to either `_SingleSampler` or `_ParallelSampler`.
+    :param dict initial_params: dict containing initial tensors to initiate
+        the markov chain. The leading dimension's size must match that of
+        `num_chains`. If not specified, parameter values will be sampled from
+        the prior.
+    :param hook_fn: Python callable that takes in `(kernel, samples, stage, i)`
+        as arguments. stage is either `sample` or `warmup` and i refers to the
+        i'th sample for the given stage. This can be
+    :param str mp_context: Multiprocessing context to use when `num_chains > 1`.
+        Only applicable for Python 3.5 and above. Use `mp_context="spawn"` for
+        CUDA.
+    :param bool disable_progbar: Disable progress bar and diagnostics update.
+    """
+    def __init__(self, kernel, num_samples, warmup_steps=None, initial_params=None,
+                 num_chains=1, hook_fn=None, mp_context=None, disable_progbar=False):
+        self.warmup_steps = num_samples if warmup_steps is None else warmup_steps  # Stan
+        self.num_samples = num_samples
+        self.kernel = kernel
+        if num_chains > 1:
+            # old API does not work since we need reference to transforms
+            # in the main process.
```

Comment:
is this issue fixed in this PR?

Reply:
This now assumes that we are passing the transforms directly to the `MCMC` class itself. But this wouldn't work if we just passed the model arg to `kernel` since with the spawn multiprocessing context, we wouldn't have access to `transforms` in the main process.